blackbird_python/blackbird/program.py

```diff
@@ -0,0 +1,290 @@
+
+        Returns:
+            int: program length
+        """
+        return len(self._operations)
+
+    def serialize(self):
+        """Serializes the blackbird program, returning a valid Blackbird script
+        as a string.
+
+        Returns:
+            str: the blackbird script representing the BlackbirdProgram object
+        """
+        # top level metadata
+        var_count = 0
+        array_insert = 3
+
+        script = ["name {}".format(self.name), "version {}".format(self.version)]
+
+        if self.target["name"] is not None:
+            array_insert += 1
+            options = ""
+
+            if self.target["options"]:
+                # if the target has options, compile them into
+                # the expected syntax
+                option_strings = [
+                    "{}={}".format(k, v)
+                    if not isinstance(v, str)
+                    else '{}="{}"'.format(k, v)
+                    for k, v in self.target["options"].items()
+                ]
+                options = " ({})".format(", ".join(option_strings))
+
+            # add target metadata
+            script.append("target {}{}".format(self.target["name"], options))
+
+        # line break
+        script.append("")
+
+
+        # loop through each quantum operation
+        for op in self.operations:
+            if len(op["modes"]) == 1:
+                modes = op["modes"][0]
+            else:
+                modes = op["modes"]
+
+            # check if the operation has any arguments
+            if "args" in op:
+                args = []
+                kwargs = []
+
+                # loop through position arguments
+                for v in op["args"]:
+                    # for each operation argument, format it
+                    # correctly depending on its type
+                    if isinstance(v, np.ndarray):
+                        # create an array variable
+                        var_name = "A{}".format(var_count)
+                        args.append(var_name)
+                        var_count += 1
+
+                        # add array declaration to script after the metadata block
+                        bb_array = numpy_to_blackbird(v, var_name)
+                        for idx, line in enumerate(bb_array):
+                            script.insert(array_insert + idx, line)
+
+                        array_insert += len(bb_array)
+
+                    elif isinstance(v, str):
+                        # argument is a string type
+                        args.append('"{}"'.format(v))
+
+                    elif isinstance(v, complex):
+                        # argument is a complex type
+                        args.append(
+                            "{}{}{}j".format(v.real, "+-"[int(v.imag < 0)], np.abs(v.imag))
+                        )
+
+                    else:
+                        args.append("{}".format(v))
```

Comment:
Which cases will this catch?

Reply:
This would catch anything that doesn't need to be dealt with as a special case, i.e. booleans, ints, floats.